blockchain/pallets/ddcclusters.go

```diff
@@ -21,26 +25,153 @@ type ClusterProps struct {
 	NodeProviderAuthContract types.AccountID
 }
 
+// Events
+type (
+	EventDdcClustersClusterCreated struct {
+		Phase     types.Phase
+		ClusterId ClusterId
+		Topics    []types.Hash
+	}
+
+	EventDdcClustersClusterNodeAdded struct {
+		Phase      types.Phase
+		ClusterId  ClusterId
+		NodePubKey NodePubKey
+		Topics     []types.Hash
+	}
+
+	EventDdcClustersClusterNodeRemoved struct {
+		Phase      types.Phase
+		ClusterId  ClusterId
+		NodePubKey NodePubKey
+		Topics     []types.Hash
+	}
+
+	EventDdcClustersClusterParamsSet struct {
+		Phase     types.Phase
+		ClusterId ClusterId
+		Topics    []types.Hash
+	}
+
+	EventDdcClustersClusterGovParamsSet struct {
+		Phase     types.Phase
+		ClusterId ClusterId
+		Topics    []types.Hash
+	}
+)
+
 type DdcClustersApi interface {
 	GetClustersNodes(clusterId ClusterId) ([]NodePubKey, error)
+	SubscribeNewClusterCreated() (*NewEventSubscription[EventDdcClustersClusterCreated], error)
```

Comment:
do we really need separate method (and duplicated implementation) for each event type? can't we have single method in BlockchainClient that allows client to specify even type he is interesting in?

e.g. blockchainClient.subscribe(ClusterNodeAdded) and receive clusterNodeAdded event channel ?

Reply:
Let's consider what `blockchainClient.Subscribe` implementation may look like.

It may accept event type instance as s parameter which should be an empty interface to allow arguments of different types.

```go
func (client *Client) Subscribe(eventT any) ???
```

We may make a type assertion in the body and return an error if the user provides an unknown type.

```go
func (client *Client) Subscribe(eventT any) (???, error) {
	switch eventT.(type) {
		case EventDdcClustersClusterNodeAdded:
			return ???, nil
		// snip
		default:
			return ???, errors.New("uknown event type")
	}
}
```

Return type should be an empty interface too because we don't know what type to return at compile type.

```go
func (client *Client) Subscribe(eventT any) (any, error) {
	switch eventT.(type) {
		case EventDdcClustersClusterNodeAdded:
			return make(chan EventDdcClustersClusterNodeAdded), nil
		// snip
		default:
			return nil, errors.New("uknown event type")
	}
}
```

This way the client will need to make a type assertion of the return value.

```go
ch, _ := client.Subscribe(EventDdcClustersClusterNodeAdded{})
ch := ch.(chan EventDdcClustersClusterNodeAdded)
```

Also no hints will be available in the client API for the user to check what subscriptions available. But the user may check available event types.

Do you see a way to avoid client-side type assertion using this approach? If no, the existing API looks better to me. What do you think?

Reply:
as for me it could be as simple as:

```
type Client struct {
        // other fields...
	eventListeners []EventListener
}

type EventListener func(events *pallets.Events)

func (c *Client) StartEventsListening() {
     // other code
     for _, eventListener := range c.eventListeners {
	 eventListener(events)
     }
     // other code
}

func main() {
	blockchainClient, _ := NewClient("")
	clusterUpdatedListener := func(events *pallets.Events) {
		for _, nodeAdded := range events.DdcClusters_ClusterNodeAdded {
			println("node added " + nodeAdded.NodePubKey.AsStoragePubKey.ToHexString())
		}
		for _, nodeDeleted := range events.DdcClusters_ClusterNodeDeleted {
			println("node deleted " + nodeDeleted.NodePubKey.AsStoragePubKey.ToHexString())
		}
	}
	blockchainClient.RegisterEventListener(clusterUpdatedListener)
}

```

so lib client (storage node) can add listeners that listen for whole `events` object and uses events which are needed. E.g. routing table can have listener that uses both DdcClusters_ClusterNodeAdded and DdcClusters_ClusterNodeDeleted events

in such implementation both SDK and client code could be simplified. 

WDYT?

Reply:
Callbacks look good, let me implement it.